"""
BacWGSpipe: a Snakemake workflow for a complete analysis of bacterial whole-genome sequencing data

BacWGSpipe supports three types of WGS data sources:
    1. Illumina short reads
    2. Long reads (either PacBio or Nanopore)
    3. Illumina short reads + Long reads (either PacBio or Nanopore) from the same isolate

Writer: Yewei Lu (lyw@cwmda.com) & Xiangchen Li (lxc@cwmda.com)
"""

##################
# Common modules #
##################
include: "rules/common.smk"

################
# Target rules #
################
# Final output of the workflow
## Modify the input section of rule all to control which part of the workflow will be run
### Hybrid mode
if config["seqdata_source"] == 0:
    # Illumina short reads + Long reads (either PacBio or Nanopore)
        rule all:
            input:
                ##### Step 01. Raw sequencing data pre-processing #####
                # Fastp QC stats
                "results/01.data_clean/short_read/all_fastp_stats.xls",
                # Kraken2 QC result
                "results/01.data_clean/short_read/all_kraken2_result.xls",
                # MultiQC report for Illumina short fastq reads QC (FastQC)
                "results/01.data_clean/short_read/multiqc/multiqc_report.html",
                # Merged NanoStat QC results for Nanopore long fastq raw reads
                "results/00.rawdata/long_read/all_raw_NanoStats.xls",
                # MultiQC report for Nanopore long fastq raw reads QC (NanoStat)
                "results/00.rawdata/long_read/multiqc_raw/multiqc_report.html",
                # Merged NanoStat QC results for Nanopore long fastq clean reads
                "results/01.data_clean/long_read/all_clean_NanoStats.xls",
                # Nanoplot report for Nanopore long fastq clean reads QC
                expand("results/01.data_clean/long_read/{sample}/nanoplot_clean/{sample}_clean_NanoPlot-report.html", sample=SAMPLES),
                # MultiQC report for Nanopore long fastq clean reads QC (NanoStat)
                "results/01.data_clean/long_read/multiqc_clean/multiqc_report.html",

                ##### Step 02. Assembly #####
                # Flag to indicate that assembly fasta files of all samples are ready
                expand("results/02.assembly/{sample}/assembly/separate_assembly.done", sample=SAMPLES),
                # Merged assembly statistics
                "results/02.assembly/all_assembly_stats.xls",
                # MulriQC report for genome assembly QC (Quast)
                "results/02.assembly/multiqc/multiqc_report.html",

                ##### Step 03. Genome annotation and Genomic component analysis #####
                # Gene length distribution plot based on prokka prediction results
                expand("results/03.genome_component/{sample}/prokka/{sample}_gene_length.png", sample=SAMPLES),
                # Merged plasmid PLSDB annotation results
                "results/03.genome_component/all_plsdb_results.xls",
                # Merged plasmid mobility prediction results generated by MOB-suite
                "results/03.genome_component/all_mob-suite_results.xls",
                # Merged Prophage sequences and annotation results generated by Phigaro
                "results/03.genome_component/all_phigaro_results.xls",
                # Merged Prediction of ICE and IME in genome sequences of bacteria for all samples by ICEfinder
                "results/03.genome_component/all_icefinder_results.xls",
                # Merged digIS summary statistics output
                "results/03.genome_component/all_digis_results.xls",
                # Merged results of genomic islands predicted by IslandPath-DIMOB
                "results/03.genome_component/all_islandpath_results.xls",
                # Merged results of CRISPRs predicted by CRISPRCasTyper
                "results/03.genome_component/all_cctyper_results.xls",

                ##### Step 04. Genome function analysis #####
                # EggNOG-mapper annotation results
                expand("results/04.gene_function/{sample}/eggnog-mapper/{sample}.emapper.annotations", sample=SAMPLES),
                # GO annotation plots
                expand("results/04.gene_function/{sample}/GO/{sample}_go.pdf", sample=SAMPLES),
                # KEGG PATHWAY annotation plots
                expand("results/04.gene_function/{sample}/KEGG/{sample}_kegg.pdf", sample=SAMPLES),
                # COG annotation plots
                expand("results/04.gene_function/{sample}/COG/{sample}_cog.pdf", sample=SAMPLES),
                # Taxonomy annotation plots
                expand("results/04.gene_function/{sample}/Taxonomy/{sample}_taxonomy.pdf", sample=SAMPLES),
                # Merged CARD annotation results
                "results/04.gene_function/all_card_results.xls",
                "results/04.gene_function/all_card_results_matrix.xls",
                "results/04.gene_function/all_card_results_matrix_binary.xlsx",
                # Merged VFDB annotation results
                "results/04.gene_function/all_vfdb_results.xls",
                "results/04.gene_function/all_vfdb_results_matrix.xls",
                "results/04.gene_function/all_vfdb_results_matrix_binary.xlsx",
                # Secretory protein results
                expand("results/04.gene_function/{sample}/secretory_protein/{sample}_secretory_protein_results.xls", sample=SAMPLES),

                ##### Step05. Genotype #####
                # Merged MLST report
                "results/05.genotype/all_mlst_results.xls",

                ##### Step 06. Genome Visualization #####
                # Flag file when the step of chromosome visualization is done
                expand("results/06.genome_visualization/{sample}/mgcplotter/mgcplotter_chr.done", sample=SAMPLES),
                # Flag file when the step of plasmid visualization is done
                expand("results/06.genome_visualization/{sample}/mgcplotter/mgcplotter_plas.done", sample=SAMPLES),

                ##### Step07. Association between AMR/VF and MGE #####
                ## AMR/VF results matrix with matched MGE results
                "results/07.association_amr_vf_mge/all_amr_mge_matrix.xls",
                "results/07.association_amr_vf_mge/all_vf_mge_matrix.xls",

                ##### Step08. Phylogenetic Analysis #####
                ## Phylogenetic tree file
                "results/08.phylogenetics/iqtree/core_gene_alignment.treefile"
### Illumina short reads ONLY
elif config["seqdata_source"] == 1:
    rule all:
        input:
            ##### Step 01. Raw sequencing data pre-processing #####
            # Fastp QC stats
            "results/01.data_clean/short_read/all_fastp_stats.xls",
            # Kraken2 QC result
            "results/01.data_clean/short_read/all_kraken2_result.xls",
            # MultiQC report for Illumina short fastq reads QC (FastQC)
            "results/01.data_clean/short_read/multiqc/multiqc_report.html",

            ##### Step 02. Assembly #####
            # Flag to indicate that assembly fasta files of all samples are ready
            expand("results/02.assembly/{sample}/assembly/separate_assembly.done", sample=SAMPLES),
            # Merged assembly statistics
            "results/02.assembly/all_assembly_stats.xls",
            # MulriQC report for genome assembly QC (Quast)
            "results/02.assembly/multiqc/multiqc_report.html",

            ##### Step 03. Genome annotation and Genomic component analysis #####
            # Gene length distribution plot based on prokka prediction results
            expand("results/03.genome_component/{sample}/prokka/{sample}_gene_length.png", sample=SAMPLES),
            # Merged results of plasmid sequences predicted by Plasmidfinder
            "results/03.genome_component/all_plasmidfinder_results.xls",
            # Merged results of plasmid sequences predicted by Platon
            "results/03.genome_component/all_platon_results.xls",
            # Merged Prophage sequences and annotation results generated by Phigaro
            "results/03.genome_component/all_phigaro_results.xls",
            # Merged Prediction of ICE and IME in genome sequences of bacteria for all samples by ICEfinder
            "results/03.genome_component/all_icefinder_results.xls",
            # Merged digIS summary statistics output
            "results/03.genome_component/all_digis_results.xls",
            # Merged results of genomic islands predicted by IslandPath-DIMOB
            "results/03.genome_component/all_islandpath_results.xls",
            # Merged results of CRISPRs predicted by CRISPRCasTyper
            "results/03.genome_component/all_cctyper_results.xls",

            ##### Step 04. Genome function analysis #####
            # EggNOG-mapper annotation results
            expand("results/04.gene_function/{sample}/eggnog-mapper/{sample}.emapper.annotations", sample=SAMPLES),
            # GO annotation plots
            expand("results/04.gene_function/{sample}/GO/{sample}_go.pdf", sample=SAMPLES),
            # KEGG PATHWAY annotation plots
            expand("results/04.gene_function/{sample}/KEGG/{sample}_kegg.pdf", sample=SAMPLES),
            # COG annotation plots
            expand("results/04.gene_function/{sample}/COG/{sample}_cog.pdf", sample=SAMPLES),
            # Taxonomy annotation plots
            expand("results/04.gene_function/{sample}/Taxonomy/{sample}_taxonomy.pdf", sample=SAMPLES),
            # Merged CARD annotation results
            "results/04.gene_function/all_card_results.xls",
            "results/04.gene_function/all_card_results_matrix.xls",
            "results/04.gene_function/all_card_results_matrix_binary.xlsx",
            # Merged VFDB annotation results
            "results/04.gene_function/all_vfdb_results.xls",
            "results/04.gene_function/all_vfdb_results_matrix.xls",
            "results/04.gene_function/all_vfdb_results_matrix_binary.xlsx",
            # Secretory protein results
            expand("results/04.gene_function/{sample}/secretory_protein/{sample}_secretory_protein_results.xls", sample=SAMPLES),

            ##### Step05. Genotype #####
            # Merged MLST report
            "results/05.genotype/all_mlst_results.xls",

            ##### Step07. Association between AMR/VF and MGE #####
            ## AMR/VF results matrix with matched MGE results
            "results/07.association_amr_vf_mge/all_amr_mge_matrix.xls",
            "results/07.association_amr_vf_mge/all_vf_mge_matrix.xls",

            ##### Step08. Phylogenetic Analysis #####
            ## Phylogenetic tree file
            "results/08.phylogenetics/iqtree/core_gene_alignment.treefile"
### Long reads (either PacBio or Nanopore) ONLY
elif config["seqdata_source"] == 2:
    rule all:
        input:
            ##### Step 01. Raw sequencing data pre-processing #####
            # Merged NanoStat QC results for Nanopore long fastq raw reads
            "results/00.rawdata/long_read/all_raw_NanoStats.xls",
            # MultiQC report for Nanopore long fastq raw reads QC (NanoStat)
            "results/00.rawdata/long_read/multiqc_raw/multiqc_report.html",
            # Merged NanoStat QC results for Nanopore long fastq clean reads
            "results/01.data_clean/long_read/all_clean_NanoStats.xls",
            # Nanoplot report for Nanopore long fastq clean reads QC
            expand("results/01.data_clean/long_read/{sample}/nanoplot_clean/{sample}_clean_NanoPlot-report.html", sample=SAMPLES),
            # MultiQC report for Nanopore long fastq clean reads QC (NanoStat)
            "results/01.data_clean/long_read/multiqc_clean/multiqc_report.html",

            ##### Step 02. Assembly #####
            # Flag to indicate that assembly fasta files of all samples are ready
            expand("results/02.assembly/{sample}/assembly/separate_assembly.done", sample=SAMPLES),
            # Merged assembly statistics
            "results/02.assembly/all_assembly_stats.xls",
            # MulriQC report for genome assembly QC (Quast)
            "results/02.assembly/multiqc/multiqc_report.html",

            ##### Step 03. Genome annotation and Genomic component analysis #####
            # Gene length distribution plot based on prokka prediction results
            expand("results/03.genome_component/{sample}/prokka/{sample}_gene_length.png", sample=SAMPLES),
            # Merged results of plasmid sequences predicted by Plasmidfinder
            "results/03.genome_component/all_plasmidfinder_results.xls",
            # Merged Prophage sequences and annotation results generated by Phigaro
            "results/03.genome_component/all_phigaro_results.xls",
            # Merged Prediction of ICE and IME in genome sequences of bacteria for all samples by ICEfinder
            "results/03.genome_component/all_icefinder_results.xls",
            # Merged digIS summary statistics output
            "results/03.genome_component/all_digis_results.xls",
            # Merged results of genomic islands predicted by IslandPath-DIMOB
            "results/03.genome_component/all_islandpath_results.xls",
            # Merged results of CRISPRs predicted by CRISPRCasTyper
            "results/03.genome_component/all_cctyper_results.xls",

            ##### Step 04. Genome function analysis #####
            # EggNOG-mapper annotation results
            expand("results/04.gene_function/{sample}/eggnog-mapper/{sample}.emapper.annotations", sample=SAMPLES),
            # GO annotation plots
            expand("results/04.gene_function/{sample}/GO/{sample}_go.pdf", sample=SAMPLES),
            # KEGG PATHWAY annotation plots
            expand("results/04.gene_function/{sample}/KEGG/{sample}_kegg.pdf", sample=SAMPLES),
            # COG annotation plots
            expand("results/04.gene_function/{sample}/COG/{sample}_cog.pdf", sample=SAMPLES),
            # Taxonomy annotation plots
            expand("results/04.gene_function/{sample}/Taxonomy/{sample}_taxonomy.pdf", sample=SAMPLES),
            # Merged CARD annotation results
            "results/04.gene_function/all_card_results.xls",
            "results/04.gene_function/all_card_results_matrix.xls",
            "results/04.gene_function/all_card_results_matrix_binary.xlsx",
            # Merged VFDB annotation results
            "results/04.gene_function/all_vfdb_results.xls",
            "results/04.gene_function/all_vfdb_results_matrix.xls",
            "results/04.gene_function/all_vfdb_results_matrix_binary.xlsx",
            # Secretory protein results
            expand("results/04.gene_function/{sample}/secretory_protein/{sample}_secretory_protein_results.xls", sample=SAMPLES),

            ##### Step05. Genotype #####
            # Merged MLST report
            "results/05.genotype/all_mlst_results.xls",

            ##### Step07. Association between AMR/VF and MGE #####
            ## AMR/VF results matrix with matched MGE results
            "results/07.association_amr_vf_mge/all_amr_mge_matrix.xls",
            "results/07.association_amr_vf_mge/all_vf_mge_matrix.xls",

            ##### Step08. Phylogenetic Analysis #####
            ## Phylogenetic tree file
            "results/08.phylogenetics/iqtree/core_gene_alignment.treefile"

###########
# Modules #
###########
include: "rules/step01.data_clean.smk"
include: "rules/step02.assembly.smk"
include: "rules/step03.genome_component.smk"
include: "rules/step04.gene_function.smk"
include: "rules/step05.genotype.smk"
include: "rules/step06.genome_visualization.smk"
include: "rules/step07.association_amr_vf_mge.smk"
include: "rules/step08.phylogenetics.smk"
